sql/ttl: restart job on node topology change

spilchen opened the pull request:
Previously, TTL jobs did not respond promptly to node failures. If a node participating in a TTL job went down, the remaining nodes would continue execution without compensating for the lost parallelism, leading to suboptimal job performance.

This change improves fault tolerance and job efficiency by introducing a mechanism to better react to topology changes. The TTL coordinator monitors for topology changes and can restart the job to leverage any newly available nodes, restoring or enhancing parallelism.

The change introduces new goroutines to control the various work units in the TTL coordinator. A summary of the goroutines involved in the TTL at the coorinadotr are as follows:

## TTL Coordinator
  - Stats collector (goroutine, preexisting)
    - Periodically collects row-level statistics for the job
  - Replan Checker (goroutine, new)
    - Periodically evaluates whether the physical plan needs replanning
    - If a significant node change is detected, the job is failed and retried
  - DistSQL execution (goroutine, new)
    - Drives overall job execution
    - Cancels the stats group and replan checker on completion
  - TTL Job Runner (control loop)
    - Waits for replan checker and DistSQL execution
    - Will log any warnings from the stats collector
    
Fixes #140513
Fixes #138999

Epic: none
Release note (performance improvement): TTL jobs now respond to node topology changes by restarting and rebalancing across available nodes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145214)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
i still need to spend more time understanding the goroutines, but i gave this an initial review. nice work so far!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-:-OPIOrqICwvS01Nk5XNk:bu65p6h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCC6Z86EHY0C3W6Bqi:-OPCC6Z86EHY0C3W6Bqj:bdgavft) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/sql/ttl/ttljob/ttljob.go#L42)):*
> ```Go
> 	"sql.ttl.replan_flow_threshold",
> 	"the fraction of flow instances that must change (added or updated) before the TTL job is replanned; set to 0 to disable",
> 	0.1,
> ```

nit: let's use the `settings.FloatInRange` validator to make sure it's between 0 and 1.
___
*[`pkg/cmd/roachtest/tests/ttl_restart.go` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCBPkwE8NJPMlZWuDS:-OPCBPkwE8NJPMlZWuDT:b-q8lajy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/cmd/roachtest/tests/ttl_restart.go#L27)):*
> ```Go
> // registerTTLRestart will register the TTL restart job. This is a test that
> // verifies the restart behavior of the TTL job. It ensures that the TTL job will
> // restart if a node goes down or a new one comes online.
> ```

nice test! do you think it would be easy to add one more variant of the test that covers this too: https://github.com/cockroachdb/cockroach/issues/138999

(or if you feel the test is already sufficient to check that, no need to do anything else)
___
*[`pkg/cmd/roachtest/tests/ttl_restart.go` line 204 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCBhYS9_DPMYrFe2BF:-OPCBhYT1JwJOLiT_P2b:b-9oozb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/cmd/roachtest/tests/ttl_restart.go#L204)):*
> ```Go
> }
> 
> func gatherLeaseDistribution(
> ```

super nit: summarizing comments on these helpers functions would be useful
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCDANm11pgdDVPNrWe:-OPCDANm11pgdDVPNrWf:b-pqe9os) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/sql/ttl/ttljob/ttljob_processor.go#L61)):*
> ```Go
> 	startAsyncWork(ctx context.Context) (completeCallback func() error, workersDoneCh chan error, cancel context.CancelFunc, retErr error)
> 	getMetadataPushInterval() time.Duration
> 	MoveToDraining(err error)
> ```

nit: some comments describing when and how these functions are called would be useful
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCDQJXBFmxFAUnqQTx:-OPCDQJXBFmxFAUnqQTy:bgvpgo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/sql/ttl/ttljob/ttljob_processor.go#L118)):*
> ```Go
> 				cancelFunc() // Stop the async work
> 			}
> 		case err = <-workersDoneCh:
> ```

do we need another case for ctx.Done()? or does this workersDone channel already include that?
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 199 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPIOZpc03wc_ri7FeJp:-OPIOZpc03wc_ri7FeJq:b-66ux1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e87a8e7c294b31455570137f1f5f97157bf8aa1/pkg/sql/ttl/ttljob/ttljob_processor.go#L199)):*
> ```Go
> 	retErr error,
> ) {
> 	ctx, cancel = context.WithCancel(ctx)
> ```

nit: let's use `context.WithCancelCause`. it lets us be more explicit about why the context is getting cancelled.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 374 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPIOWjwEu7ytRJcGSlg:-OPIOWjwEu7ytRJcGSlh:b-j6cdy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e87a8e7c294b31455570137f1f5f97157bf8aa1/pkg/sql/ttl/ttljob/ttljob_processor.go#L374)):*
> ```Go
> 			case <-ctx.Done():
> 				log.VInfof(ctx, 2 /* level */, "stopping TTL processor bounds producer: %s", ctx.Err())
> 				return errors.Wrapf(ctx.Err(), "stopping TTL processor bounds producer")
> ```

nit: let's log this at the normal level; it shouldn't happen in most cases
nit: we can use `context.Cause(ctx)` to get a better error message in some cases.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 411 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPINNnTC66E3ChyG4DB:-OPINNnTC66E3ChyG4DC:b-roxsyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e87a8e7c294b31455570137f1f5f97157bf8aa1/pkg/sql/ttl/ttljob/ttljob_processor.go#L411)):*
> ```Go
> 
> 	workersDoneCh = make(chan error)
> 	go func() {
> ```

to avoid having this goroutine interfere with server shutdown, we should use `stopper.RunAsyncTask` here.... though i'm not sure if we have access to the stopper. perhaps this is fine if we know that the ctxgroup will always exit during shutdown.

all that being said, can you add a comment here saying why we need this channel?
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 504 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPIOFB87s_mPYgMNmwW:-OPIOFB87s_mPYgMNmwX:b-73h4bc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e87a8e7c294b31455570137f1f5f97157bf8aa1/pkg/sql/ttl/ttljob/ttljob_processor.go#L504)):*
> ```Go
> 		case <-ctx.Done():
> 			log.VInfof(ctx, 2 /* level */, "stopping TTL processor on query bounds: %s", ctx.Err())
> 			return spanRowCount, errors.Wrapf(ctx.Err(), "stopping TTL processor on query bounds")
> ```

ditto above about the cause and log level


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145214)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/cmd/roachtest/tests/ttl_restart.go` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCBPkwE8NJPMlZWuDS:-OPptbxQETy-6b9KZm5y:b-rbn35p) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/cmd/roachtest/tests/ttl_restart.go#L27)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nice test! do you think it would be easy to add one more variant of the test that covers this too: https://github.com/cockroachdb/cockroach/issues/138999

(or if you feel the test is already sufficient to check that, no need to do anything else)
</blockquote></details>

Good idea. I'll include a variant for that.
___
*[`pkg/cmd/roachtest/tests/ttl_restart.go` line 204 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCBhYS9_DPMYrFe2BF:-OPpnpej-8LX16177kdJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/cmd/roachtest/tests/ttl_restart.go#L204)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: summarizing comments on these helpers functions would be useful
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPCDQJXBFmxFAUnqQTx:-OPpXQBk4YDPyTsQHOMh:bsevkji) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f850e9f10a90b63de4288c66d99ad8b8caef0a9/pkg/sql/ttl/ttljob/ttljob_processor.go#L118)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do we need another case for ctx.Done()? or does this workersDone channel already include that?
</blockquote></details>

It's already handled, though indirectly. We check `ctx.Done()` within the low-level goroutines started by `startAsyncWork`. When the context is canceled, it triggers an error that gets published to the `workersDoneCh` channel., which will find its way here. I admit it isn't simple. Go doesn't seem to have an easy way to wait for a group of goroutines in an async manner.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 411 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPINNnTC66E3ChyG4DB:-OPpfO4H8LnfYWL8da7l:bjapij) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e87a8e7c294b31455570137f1f5f97157bf8aa1/pkg/sql/ttl/ttljob/ttljob_processor.go#L411)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

to avoid having this goroutine interfere with server shutdown, we should use `stopper.RunAsyncTask` here.... though i'm not sure if we have access to the stopper. perhaps this is fine if we know that the ctxgroup will always exit during shutdown.

all that being said, can you add a comment here saying why we need this channel?
</blockquote></details>

I don't think `stopper` is available to use, and passing it down to the processor wouldn't be ideal. I'll add a comment.

```
	// workersDoneCh signals when the above processor goroutines have exited.
	// This allows the caller to poll on this channel to detect completion in a
	// non-blocking way.
```

Note, the ttl_restart test does heavy testing of stopping a node while TTL is running. So, I think that part is well tested.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 504 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OPIOFB87s_mPYgMNmwW:-OPpmzb35uI8qJKEu0dy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e87a8e7c294b31455570137f1f5f97157bf8aa1/pkg/sql/ttl/ttljob/ttljob_processor.go#L504)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ditto above about the cause and log level
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145214)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 284 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OQQBkd2AodQzw51qBSG:-OQQBkd2AodQzw51qBSH:blfxut2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb724d6616a5325b9a54edc1177fff5c7e23729/pkg/sql/ttl/ttljob/ttljob.go#L284)):*
> ```Go
> 	// there are idle nodes that become available, it's more efficient to restart
> 	// the TTL job to utilize those nodes for parallel work.
> 	replanChecker, cancelReplanner := sql.PhysicalPlanChangeChecker(
> ```

for backport safety, i think it would be worth it for us to test how much of an improvement we get _just_ by adding this `PhysicalPlanChangeChecker`. if that helps on its own, we should consider making a commit that just adds the replanChecker and makes no other changes.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OQQDYaEBW8jTAgd5wFN:-OQQDYaEBW8jTAgd5wFO:b-ts4l8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb724d6616a5325b9a54edc1177fff5c7e23729/pkg/sql/ttl/ttljob/ttljob_processor.go#L125)):*
> ```Go
> 			// notification if another node goes down, preventing the processor from
> 			// completing all spans before realizing the failure.
> 			if status := output.Push(nil, &execinfrapb.ProducerMetadata{}); status != execinfra.NeedMoreRows {
> ```

ultimately, our goal here is to resolve https://github.com/cockroachdb/cockroach/issues/135227. to do that, we will eventually need to change this part so that it pushes back a progress update rather than an empty metadata.

that means we would need to make this main goroutine receive progress updates from the ones started in `startAsyncWork`, which seems like it could get complex.

i'm wondering if a better design would be to keep the "feed query bounds" work inside of this main goroutine. we could still have it factored into a separate function for legibility, but at a high level the only other change would be that this select block would need another case:
```

// call the function that calculates the query bounds

for i, span := range spans {
  select {
  case <-tick.C:
    // keep sending an empty metadata here to detect failures.
  case boundsChan <- bounds:
    // push a progress update metadata here.
    // reset the ticker.
  case err = <-workersDoneCh:
    // unchanged
  }
}
```

actually we might even be able to make it even simpler -- we shouldn't really need the `workersDoneCh` any more, since the for loop will end after we iterate over all of the `range spans`. after the for loop has ended, we can just make one blocking call to `g.Wait()`
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 201 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OQQCjpPEjoAXbiGQvnz:-OQQCjpQ1Wq22L0iK39n:b89gnjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb724d6616a5325b9a54edc1177fff5c7e23729/pkg/sql/ttl/ttljob/ttljob_processor.go#L201)):*
> ```Go
> ) (
> 	completeCallback func() error,
> 	workersDoneCh chan error,
> ```

super nit: if we make this of type `<-chan`, then it wlll be more clear that the returned channel is only intended to be read from. (similar to the return value of `ctx.Done()`


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145214)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 284 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OQQBkd2AodQzw51qBSG:-OQiTxKQ94pJfkVSfKLc:byt4vd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb724d6616a5325b9a54edc1177fff5c7e23729/pkg/sql/ttl/ttljob/ttljob.go#L284)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

for backport safety, i think it would be worth it for us to test how much of an improvement we get _just_ by adding this `PhysicalPlanChangeChecker`. if that helps on its own, we should consider making a commit that just adds the replanChecker and makes no other changes.
</blockquote></details>

I found that just having the plan checker here resolved the issue. I was able to see that the TTL job restarts if node membership changes. So, I will proceed with a stripped down version of this changes -- essentially most of the code in ttljob_processor.go will be removed.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OQQDYaEBW8jTAgd5wFN:-OQiUlgh2FCmZpuvLqUG:b-wt7xy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb724d6616a5325b9a54edc1177fff5c7e23729/pkg/sql/ttl/ttljob/ttljob_processor.go#L125)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ultimately, our goal here is to resolve https://github.com/cockroachdb/cockroach/issues/135227. to do that, we will eventually need to change this part so that it pushes back a progress update rather than an empty metadata.

that means we would need to make this main goroutine receive progress updates from the ones started in `startAsyncWork`, which seems like it could get complex.

i'm wondering if a better design would be to keep the "feed query bounds" work inside of this main goroutine. we could still have it factored into a separate function for legibility, but at a high level the only other change would be that this select block would need another case:
```

// call the function that calculates the query bounds

for i, span := range spans {
  select {
  case <-tick.C:
    // keep sending an empty metadata here to detect failures.
  case boundsChan <- bounds:
    // push a progress update metadata here.
    // reset the ticker.
  case err = <-workersDoneCh:
    // unchanged
  }
}
```

actually we might even be able to make it even simpler -- we shouldn't really need the `workersDoneCh` any more, since the for loop will end after we iterate over all of the `range spans`. after the for loop has ended, we can just make one blocking call to `g.Wait()`
</blockquote></details>

I'll include this in the follow-on PR, since most of the code in `ttljob_processor.go` was backed out.

On the topic of simplification: I think we still need to wait in a non-blocking way. Otherwise, we risk halting progress updates to the coordinator. My understanding is that progress is reported as the select-delete goroutines process spans. If the main control loop is blocked on `g.Wait()`, then no updates would be sent until all goroutines complete. Ideally, we should be sending updates as each goroutine finishes its span.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 201 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OQQCjpPEjoAXbiGQvnz:-OQiWGrpBlKYgpqBTGZ9:bkomni8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb724d6616a5325b9a54edc1177fff5c7e23729/pkg/sql/ttl/ttljob/ttljob_processor.go#L201)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: if we make this of type `<-chan`, then it wlll be more clear that the returned channel is only intended to be read from. (similar to the return value of `ctx.Done()`
</blockquote></details>

I'll keep this in mind when I work on the follow-on PR.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145214)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-OQQDYaEBW8jTAgd5wFN:-OQicA302xWLIBjyS54C:bs96pay) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bb724d6616a5325b9a54edc1177fff5c7e23729/pkg/sql/ttl/ttljob/ttljob_processor.go#L125)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I'll include this in the follow-on PR, since most of the code in `ttljob_processor.go` was backed out.

On the topic of simplification: I think we still need to wait in a non-blocking way. Otherwise, we risk halting progress updates to the coordinator. My understanding is that progress is reported as the select-delete goroutines process spans. If the main control loop is blocked on `g.Wait()`, then no updates would be sent until all goroutines complete. Ideally, we should be sending updates as each goroutine finishes its span.
</blockquote></details>

sounds good. we can discuss further on the next PR, but my idea was to call `g.Wait()` only _after_ the for loop over `range spans` completes. 

i think i see your point though. maybe the way around that is to have another channel named `progressCh`, and each worker can push progress into that channel as they process a span. (basically, we would have this `progressCh` and read from it in the main control loop, instead of reading once from `workersDoneCh`.)


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/ttl: restart job on node topology change

Previously, TTL jobs did not respond promptly to node failures. If a
node participating in a TTL job went down, the remaining nodes would
continue execution without compensating for the lost parallelism,
leading to suboptimal job performance.

This change improves fault tolerance and job efficiency by introducing
a mechanisms to better react to topology changes. The TTL coordinator
monitors for topology changes and can restart the job to leverage any
newly available nodes, restoring or enhancing parallelism.

The change introduces new goroutines to control the various work units
in the TTL coordinator.  A summary of the goroutines involved in the TTL
at the coorindator are as follows:

TTL Coordinator
  - Stats collector (goroutine, preexisting)
    - Periodically collects row-level statistics for the job
  - Replan Checker (goroutine, new)
    - Periodically evaluates whether the physical plan needs replanning
    - If a significant node change is detected, the job is failed and retried
  - DistSQL execution (goroutine, new)
    - Drives overall job execution
    - Cancels the stats group and replan checker on completion
  - TTL Job Runner (control loop)
    - Waits for replan checker and DistSQL execution
    - Will log any warnings from the stats collector

Fixes #140513

Epic: none
Release note (performance improvement): TTL jobs now respond to node
topology changes by restarting and rebalancing across available nodes.

rafiss reviewed (approved):
lgtm! thanks for doing the work to test with the simplified PR, and splitting off the processor changes!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145214#-:-OQnQ8K_65r0FrvWlIjq:b-bmdto4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
